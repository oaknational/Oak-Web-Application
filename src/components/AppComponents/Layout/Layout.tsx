"use client";
import Head from "next/head";
import React, { FC } from "react";
import { useRouter } from "next/navigation";
import {
  OakBox,
  OakFlex,
  oakDefaultTheme,
  OakThemeProvider,
  OakColorToken,
} from "@oaknational/oak-components";

import Seo, { SeoProps } from "@/browser-lib/seo/Seo";
import AppHeader from "@/components/AppComponents/AppHeader";
import LayoutSiteFooter from "@/components/AppComponents/LayoutSiteFooter";
import { OrganizationJsonLd } from "@/browser-lib/seo/getJsonLd";
import LayoutPreviewControls from "@/components/AppComponents/LayoutPreviewControls";
import LayoutClientErrorHeader from "@/components/AppComponents/LayoutClientErrorHeader";
import LayoutClientErrorFooter from "@/components/AppComponents/LayoutClientErrorFooter";
import LandingPagesHeader, {
  LayoutLandingPagesHeaderProps,
} from "@/components/AppComponents/LayoutLandingPagesHeader";
import { CTA } from "@/common-lib/cms-types";
import { Breadcrumb } from "@/components/SharedComponents/Breadcrumbs";
import SkipLink from "@/components/CurriculumComponents/OakComponentsKitchen/SkipLink";

export type HeaderProps = {
  children?: React.ReactNode;
  breadcrumbs?: Breadcrumb[];
};

export type HeaderVariant = "app" | "landing-pages" | "client-error";
const headers: Record<
  HeaderVariant,
  FC | FC<LayoutLandingPagesHeaderProps> | FC<HeaderProps>
> = {
  app: AppHeader,
  "landing-pages": LandingPagesHeader,
  "client-error": LayoutClientErrorHeader,
};

export type FooterVariant = "default" | "client-error";
const footers: Record<FooterVariant, FC> = {
  default: LayoutSiteFooter,
  "client-error": LayoutClientErrorFooter,
};

export type LayoutProps = {
  children?: React.ReactNode;
  seoProps: SeoProps;
  headerVariant?: HeaderVariant;
  footerVariant?: FooterVariant;
  breadcrumbs?: Breadcrumb[];
  $background?: OakColorToken;
  headerCta?: CTA | null;
  banner?: React.ReactNode;
};

const Layout: FC<LayoutProps> = (props) => {
  const {
    children,
    seoProps,
    $background,
    breadcrumbs,
    headerVariant = "app",
    footerVariant = "default",
    banner,
  } = props;
  const Header = headers[headerVariant];
  const Footer = footers[footerVariant];
  // TODO: draft mode
  // const { isPreview } = useRouter();

  return (
    <>
      <OakThemeProvider theme={oakDefaultTheme}>
        <Seo {...seoProps} />
        <Head>
          <link rel="icon" href="/favicon.ico" />
        </Head>
        <OrganizationJsonLd />
        <OakFlex
          $flexDirection="column"
          $flexGrow={1}
          $background={$background}
        >
          <OakBox
            $position={"absolute"}
            $height={"all-spacing-13"}
            $width={"all-spacing-11"}
            $zIndex={"in-front"}
            $top={"all-spacing-14"}
            $left={"all-spacing-6"}
          >
            <SkipLink href="#main">Skip to content</SkipLink>
          </OakBox>
          {banner}
          <Header breadcrumbs={breadcrumbs} headerCta={props.headerCta} />
          <OakFlex
            $flexDirection="column"
            $flexGrow={1}
            $width="100%"
            as="main"
            id="main"
          >
            {children}
          </OakFlex>
          <Footer />
          {/* {isPreview && <LayoutPreviewControls />} */}
        </OakFlex>
      </OakThemeProvider>
    </>
  );
};

export default Layout;
