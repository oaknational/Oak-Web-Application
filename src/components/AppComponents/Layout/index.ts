export { default } from "./Layout";
export type { LayoutProps } from "./Layout";
