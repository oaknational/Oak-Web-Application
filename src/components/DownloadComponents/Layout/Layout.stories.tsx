import type { Meta, StoryObj } from "@storybook/react";
import React from "react";

import Component from "./Layout";

const meta: Meta<typeof Component> = {
  title: "Download Components/Layout",
  component: Component,
};

export default meta;
type Story = StoryObj<typeof Component>;

export const Layout: Story = {
  args: {
    header: "Lesson resources",
    cardGroup: <div>Mock Card Group</div>,
    ctaButton: <button>CTA</button>,
    userDetails: <div>Mock user details</div>,
  },
  render: (args) => {
    return <Component {...args} />;
  },
};
