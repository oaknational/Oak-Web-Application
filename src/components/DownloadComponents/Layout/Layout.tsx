import { FC } from "react";

import { Heading } from "@/components/Typography";
import FieldError from "@/components/FormFields/FieldError";
import Box from "@/components/Box";
import Checkbox from "@/components/Checkbox";
import Flex from "@/components/Flex";

/** Generic layout component for Downloads and Share page */

type LayoutProps = {
  header: string;
  handleToggle: () => void;
  selectAllChecked: boolean;
  errorMessage?: string;
  cardGroup: React.ReactNode;
  userDetails: React.ReactNode;
  ctaButton: React.ReactNode;
};

const Layout: FC<LayoutProps> = (props) => {
  return (
    <Box $width="100%">
      <Flex $alignItems={"flex-start"} $flexDirection={"column"} $gap={32}>
        <Heading tag="h1" $font={["heading-5", "heading-4"]}>
          {props.header}
        </Heading>
        <Flex $justifyContent="space-between" $width="100%">
          <Flex $flexDirection="column" $gap={24}>
            <Heading tag="h3" $font={["heading-6", "heading-5"]}>
              Lesson Resources
            </Heading>
            <Box $maxWidth="max-content">
              <Checkbox
                checked={props.selectAllChecked}
                onChange={props.handleToggle}
                id="select-all"
                name="select-all"
                variant="withLabel"
                labelText="Select all"
                labelFontWeight={600}
              />
            </Box>
            <FieldError id={"downloads-error"}>{props.errorMessage}</FieldError>
            {props.cardGroup}
          </Flex>
          <Flex $flexDirection="column">
            {props.userDetails}
            {props.ctaButton}
          </Flex>
        </Flex>
      </Flex>
    </Box>
  );
};

export default Layout;
