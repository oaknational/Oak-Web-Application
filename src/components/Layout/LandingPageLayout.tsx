import Head from "next/head";
import { FC } from "react";
import styled from "styled-components";

import SiteHeader from "../SiteHeader";
import SiteFooter from "../SiteFooter";
import { OrganizationJsonLd } from "../../browser-lib/seo/getJsonLd";
import Seo, { SeoProps } from "../../browser-lib/seo/Seo";
import background, { BackgroundProps } from "../../styles/utils/background";
import { OakColorName } from "../../styles/theme";

const Container = styled.div<BackgroundProps>`
  display: flex;
  flex-direction: column;
  flex-grow: 1;
  ${background}
`;
const StyledLayout = styled.main`
  display: flex;
  flex-direction: column;
  flex-grow: 1;
  padding: 0 12px;
  max-width: 1200px;
  width: 100%;
  align-self: center;
`;

interface LayoutProps {
  seoProps: SeoProps;
  background?: OakColorName;
}

/** 1. Titles for SEO should be between 50-60 characters long 
    2. Title should contain app name
    3. SEO descriptions should be between 150-300 characters long */
const LandingPageLayout: FC<LayoutProps> = (props) => {
  const { children, seoProps, background } = props;
  return (
    <>
      <Seo {...seoProps} />
      <Head>
        <link rel="icon" href="/favicon.ico" />
      </Head>
      <OrganizationJsonLd />
      <Container background={background}>
        <SiteHeader />
        <StyledLayout>{children}</StyledLayout>
        <SiteFooter />
      </Container>
    </>
  );
};

export default LandingPageLayout;
