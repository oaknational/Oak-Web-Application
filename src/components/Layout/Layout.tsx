import Head from "next/head";
import { FC } from "react";
import { useRouter } from "next/router";
import styled from "styled-components";

import AppHeader from "../AppHeader";
import SiteFooter from "../SiteFooter";
import { OrganizationJsonLd } from "../../browser-lib/seo/getJsonLd";
import Seo, { SeoProps } from "../../browser-lib/seo/Seo";
import background, { BackgroundProps } from "../../styles/utils/background";
import { OakColorName } from "../../styles/theme";
import SiteHeader from "../SiteHeader";
import PreviewControls from "../PreviewControls";
import ClientErrorHeader from "../ClientErrorHeader";
import ClientErrorFooter from "../ClientErrorFooter";
import LandingPagesHeader from "../LandingPagesHeader";
import { CTA } from "../../node-lib/cms";
import { LandingPagesHeaderProps } from "../LandingPagesHeader/LandingPagesHeader";
import { Breadcrumb } from "../Breadcrumbs";

const Container = styled.div<BackgroundProps>`
  display: flex;
  flex-direction: column;
  flex-grow: 1;
  ${background}
`;
const StyledLayout = styled.main`
  display: flex;
  flex-direction: column;
  flex-grow: 1;
  width: 100%;
`;

export type HeaderProps = {
  breadcrumbs?: Breadcrumb[];
};

export type HeaderVariant = "app" | "site" | "landing-pages" | "client-error";
const headers: Record<
  HeaderVariant,
  FC | FC<LandingPagesHeaderProps> | FC<HeaderProps>
> = {
  app: AppHeader,
  site: SiteHeader,
  "landing-pages": LandingPagesHeader,
  "client-error": ClientErrorHeader,
};

export type FooterVariant = "default" | "client-error";
const footers: Record<FooterVariant, FC> = {
  default: SiteFooter,
  "client-error": ClientErrorFooter,
};

export type LayoutProps = {
  seoProps: SeoProps;
  headerVariant?: HeaderVariant;
  footerVariant?: FooterVariant;
  breadcrumbs?: Breadcrumb[];
  $background?: OakColorName;
  headerCta?: CTA | null;
};

const Layout: FC<LayoutProps> = (props) => {
  const {
    children,
    seoProps,
    $background,
    breadcrumbs,
    headerVariant = "site",
    footerVariant = "default",
  } = props;
  const Header = headers[headerVariant];
  const Footer = footers[footerVariant];

  const { isPreview } = useRouter();

  return (
    <>
      <Seo {...seoProps} />
      <Head>
        <link rel="icon" href="/favicon.ico" />
      </Head>
      <OrganizationJsonLd />
      <Container $background={$background}>
        <Header breadcrumbs={breadcrumbs} headerCta={props.headerCta} />

        <StyledLayout>{children}</StyledLayout>
        <Footer />
        {isPreview && <PreviewControls />}
      </Container>
    </>
  );
};

export default Layout;
