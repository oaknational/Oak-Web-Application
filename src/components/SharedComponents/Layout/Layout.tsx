import Head from "next/head";
import React, { FC } from "react";
import { useRouter } from "next/router";
import styled from "styled-components";

import Seo, { SeoProps } from "@/browser-lib/seo/Seo";
import AppHeader from "@/components/AppComponents/AppHeader";
import SiteFooter from "@/components/SiteFooter";
import { OrganizationJsonLd } from "@/browser-lib/seo/getJsonLd";
import background, { BackgroundProps } from "@/styles/utils/background";
import { OakColorName } from "@/styles/theme";
import PreviewControls from "@/components/PreviewControls";
import LayoutClientErrorHeader from "@/components/SharedComponents/LayoutClientErrorHeader";
import LayoutClientErrorFooter from "@/components/SharedComponents/LayoutClientErrorFooter";
import LandingPagesHeader from "@/components/LandingPagesHeader";
import { CTA } from "@/common-lib/cms-types";
import { LandingPagesHeaderProps } from "@/components/LandingPagesHeader/LandingPagesHeader";
import { Breadcrumb } from "@/components/SharedComponents/Breadcrumbs";

const Container = styled.div<BackgroundProps>`
  display: flex;
  flex-direction: column;
  flex-grow: 1;
  ${background}
`;
const StyledLayout = styled.main`
  display: flex;
  flex-direction: column;
  flex-grow: 1;
  width: 100%;
`;

export type HeaderProps = {
  children?: React.ReactNode;
  breadcrumbs?: Breadcrumb[];
};

export type HeaderVariant = "app" | "landing-pages" | "client-error";
const headers: Record<
  HeaderVariant,
  FC | FC<LandingPagesHeaderProps> | FC<HeaderProps>
> = {
  app: AppHeader,
  "landing-pages": LandingPagesHeader,
  "client-error": LayoutClientErrorHeader,
};

export type FooterVariant = "default" | "client-error";
const footers: Record<FooterVariant, FC> = {
  default: SiteFooter,
  "client-error": LayoutClientErrorFooter,
};

export type LayoutProps = {
  children?: React.ReactNode;
  seoProps: SeoProps;
  headerVariant?: HeaderVariant;
  footerVariant?: FooterVariant;
  breadcrumbs?: Breadcrumb[];
  $background?: OakColorName;
  headerCta?: CTA | null;
  banner?: React.ReactNode;
};

const Layout: FC<LayoutProps> = (props) => {
  const {
    children,
    seoProps,
    $background,
    breadcrumbs,
    headerVariant = "app",
    footerVariant = "default",
    banner,
  } = props;
  const Header = headers[headerVariant];
  const Footer = footers[footerVariant];

  const { isPreview } = useRouter();

  return (
    <>
      <Seo {...seoProps} />
      <Head>
        <link rel="icon" href="/favicon.ico" />
      </Head>
      <OrganizationJsonLd />
      <Container $background={$background}>
        {banner}
        <Header breadcrumbs={breadcrumbs} headerCta={props.headerCta} />
        <StyledLayout>{children}</StyledLayout>
        <Footer />
        {isPreview && <PreviewControls />}
      </Container>
    </>
  );
};

export default Layout;
