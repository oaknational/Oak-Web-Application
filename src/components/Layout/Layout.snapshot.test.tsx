import renderWithProviders from "../../__tests__/__helpers__/renderWithProviders";

import Layout from "./Layout";

it("renders <Layout> unchanged", () => {
  const { container } = renderWithProviders(<Layout />);
  expect(container).toMatchSnapshot();
});
