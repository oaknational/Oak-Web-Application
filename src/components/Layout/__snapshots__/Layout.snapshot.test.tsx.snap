// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders <Layout> unchanged 1`] = `
<div>
  <header
    class="header"
  >
    <div
      class="title"
    >
      Oak
    </div>
    <input
      class="search-input"
      placeholder="Search"
      type="text"
    />
    <a
      href="/sign-in"
    >
      Sign in
    </a>
  </header>
  <main
    class="main"
  />
  <footer
    class="footer"
  >
    Â© Oak National Academy 
    2022
  </footer>
</div>
`;
